### PR TITLE
[WIP] QJson

### PR DESCRIPTION
**Pull request by elpaso**
Slightly faster JSON generation.

How much faster?

```
PASS   : TestQgsJsonUtils::testExportAttributesQJson(Use V2 (QJson))
RESULT : TestQgsJsonUtils::testExportAttributesQJson():"Use V2 (QJson)":
     0.0048 msecs per iteration (total: 79, iterations: 16384)
PASS   : TestQgsJsonUtils::testExportAttributesQJson(Use old string concat)
RESULT : TestQgsJsonUtils::testExportAttributesQJson():"Use old string concat":
     0.0072 msecs per iteration (total: 59, iterations: 8192)
PASS   : TestQgsJsonUtils::testExportFeatureQJson(Use V2 (QJson))
RESULT : TestQgsJsonUtils::testExportFeatureQJson():"Use V2 (QJson)":
     0.041 msecs per iteration (total: 84, iterations: 2048)
PASS   : TestQgsJsonUtils::testExportFeatureQJson(Use old string concat)
RESULT : TestQgsJsonUtils::testExportFeatureQJson():"Use old string concat":
     0.046 msecs per iteration (total: 95, iterations: 2048)

```

Btw, the main advantage is that we are not building JSON by concatenating strings, which is error prone.

**Funded by: QCooperative.net**

### CODE REVIEW

File: src/core/qgsjsonutils.h [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include "qgsfields.h"
 
 #include <QPointer>
+#include <QJsonValue>
```
Comment (m-kuhn) [inline src/core/qgsjsonutils.h]:
```suggestion
#include <QJsonObject>
```

File: src/core/qgsjsonutils.h [hunk 2]
```diff
@@ -198,6 +199,19 @@ class CORE_EXPORT QgsJsonExporter
                            const QVariantMap &extraProperties = QVariantMap(),
                            const QVariant &id = QVariant() ) const;
 
+    /**
+     * Returns a GeoJSON string representation of a feature.
```
Comment (m-kuhn) [inline src/core/qgsjsonutils.h]: Is it returning a string or a QJsonObject (or is the QJSonObject a string)?

Comment (elpaso) [reply src/core/qgsjsonutils.h]: @m-kuhn please wait, this is just POC, I didn't fix all docstrings and tests are missing, I'm just looking for general feedback on the idea.

**Comment by m-kuhn** [comment]
Good work!
Should we have a V2 prefix or simply call them `toJsonObject` instead?

File: src/core/qgsjsonutils.cpp [hunk 3]
```diff
@@ -235,14 +235,143 @@ QString QgsJsonExporter::exportFeature( const QgsFeature &feature, const QVarian
...
+    {
+      QgsFields fields = mLayer ? mLayer->fields() : feature.fields();
+      // List of formatters through we want to pass the values
+      QStringList formattersWhiteList;
+      formattersWhiteList << QStringLiteral( "KeyValue" )
+                          << QStringLiteral( "List" )
+                          << QStringLiteral( "ValueRelation" )
+                          << QStringLiteral( "ValueMap" );
+
+      for ( int i = 0; i < fields.count(); ++i )
```
Comment (m-kuhn) [inline src/core/qgsjsonutils.cpp]: Could this be a range-based for loop?

**Comment by elpaso** [comment]
> Good work!
> Should we have a V2 prefix or simply call them `toJsonObject` instead?

That's the feed-back I need :) 

I like `toJsonObject`, I think that the string-returning methods will be eventually become just wrappers around the QJson-returning methods. 

The only missing bit is the `precision` argument, AFAIK there is no way to pass that into QJson* methods.

If we go that route,  string-returning methods will be eventually deprecated (because of lack of precision support).

**Comment by nyalldawson** [comment]
Couldn't the precision setting be reproduced by rounding the coordinates in QgsGeometryUtils::pointsToJsonV2?

**Comment by elpaso** [comment]
@nyalldawson good idea, I'll try that.

File: src/core/qgsjsonutils.h (hunk 2 above)
Comment (m-kuhn) [reply src/core/qgsjsonutils.h]: @elpaso all good, no hurry. There are two reasons that make me comment (likely to be continued) POC's anyway: 1. once I look at the code and see things, I don't want to care for it a second time. 2. Sometimes POC's are suddenly "accepted and merged" in less than 24 h.

Comment (elpaso) [reply src/core/qgsjsonutils.h]: This one is falling down in my TODO list btw ...

File: python/core/auto_generated/geometry/qgscircularstring.sip.in [hunk 4]
```diff
@@ -81,6 +81,7 @@ to ``p2`` will be used (i.e. winding the other way around the circle).
 
     virtual QString asJson( int precision = 17 ) const;
 
+    virtual QJsonObject asJsonObject( int precision ) const;
```
Comment (m-kuhn) [inline python/core/auto_generated/geometry/qgscircularstring.sip.in]: Should this have a default value?

**Comment by m-kuhn** [comment]
Looking great!

File: python/core/auto_generated/geometry/qgscircularstring.sip.in (hunk 4 above)
Comment (elpaso) [reply python/core/auto_generated/geometry/qgscircularstring.sip.in]: Yes, thanks for finding this.

**Comment by elpaso** [comment]
I've run the tests on a release build and `QJson` is 25% slower than the string concatenation, I'm now testing other solutions (because string concat is **NOT** the way to go).

**Pull request closed without merging**